pkg/resources/schemas/template.go

```diff
@@ -42,82 +45,108 @@ type Store struct {
+// Watch will return a APIevent channel that tracks changes to schemas for a user in a given APIRequest.
+// Changes will be returned until Done is closed on the context in the given APIRequest.
+func (s *Store) Watch(apiOp *types.APIRequest, _ *types.APISchema, _ types.WatchRequest) (chan types.APIEvent, error) {
 	user, ok := request.UserFrom(apiOp.Request.Context())
 	if !ok {
 		return nil, validation.Unauthorized
 	}
 
 	wg := sync.WaitGroup{}
-	wg.Add(2)
+	wg.Add(1)
 	result := make(chan types.APIEvent)
 
 	go func() {
 		wg.Wait()
 		close(result)
 	}()
 
-	go func() {
-		defer wg.Done()
-		c, err := s.schemaChangeNotify(apiOp.Context())
-		if err != nil {
-			return
-		}
-		schemas, err := s.sf.Schemas(user)
-		if err != nil {
-			logrus.Errorf("failed to generate schemas for user %v: %v", user, err)
-			return
-		}
-		for range c {
-			schemas = s.sendSchemas(result, apiOp, user, schemas)
-		}
-	}()
+	schemas, err := s.sf.Schemas(user)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate schemas for user '%v': %w", user, err)
+	}
+
+	// Create child contexts that allows us to cancel both change notifications routines.
+	notifyCtx, notifyCancel := context.WithCancel(apiOp.Context())
+
+	schemaChangeSignal, err := s.schemaChangeNotify(notifyCtx)
+	if err != nil {
+		notifyCancel()
+		return nil, fmt.Errorf("failed to start schema change notifications: %w", err)
+	}
+
+	userChangeSignal := s.userChangeNotify(notifyCtx, user)
 
 	go func() {
+		defer notifyCancel()
 		defer wg.Done()
-		schemas, err := s.sf.Schemas(user)
-		if err != nil {
-			logrus.Errorf("failed to generate schemas for notify user %v: %v", user, err)
-			return
-		}
-		for range s.userChangeNotify(apiOp.Context(), user) {
+
+		// For each change notification send schema updates onto the result channel.
+		for {
+			select {
+			case _, ok := <-schemaChangeSignal:
+				if !ok {
+					return
+				}
+			case _, ok := <-userChangeSignal:
+				if !ok {
+					return
+				}
+			}
 			schemas = s.sendSchemas(result, apiOp, user, schemas)
 		}
 	}()
 
 	return result, nil
 }
 
+// sendSchemas will send APIEvents onto the provided result channel based on detected changes in the schemas for the provided users.
 func (s *Store) sendSchemas(result chan types.APIEvent, apiOp *types.APIRequest, user user.Info, oldSchemas *types.APISchemas) *types.APISchemas {
+	// get the current schemas for a user
 	schemas, err := s.sf.Schemas(user)
 	if err != nil {
 		logrus.Errorf("failed to get schemas for %v: %v", user, err)
 		return oldSchemas
 	}
 
 	inNewSchemas := map[string]bool{}
-	for _, apiObject := range schemastore.FilterSchemas(apiOp, schemas.Schemas).Objects {
+
+	// Convert the schemas for the given user to a flat list of APIObjects.
+	apiObjects := schemastore.FilterSchemas(apiOp, schemas.Schemas).Objects
+	for i := range apiObjects {
+		apiObject := apiObjects[i]
```

Comment:
This might be a nitpick but wouldn't it be more idiomatic to
```suggestion
	for _, apiObject := range apiObjects {
```

Reply:
The suggested way is easier to read, but some scenarios cause unintended behavior. For instance, in this case, `apiObjects` is a list of `types.APIObject` not `*types.APIObject`, so every iteration using the suggested range format would make a new shallow copy at the start of the loop of `types.APIObject` instead of just referencing the existing object.

If you are interested, there is currently some discussion of changing the behavior of range loops because of these idiosyncrasies. 
https://github.com/golang/go/discussions/56010